### DISCUSSION TITLE
Transaction race conditions in pb_hooks?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#4125] [Q&A] [answered]
This is a bit of along shot, but I'm starting to wonder if I'm experiencing a race condition of some sort.  Will this deadlock?

```
const foo = ()=>{
   // do stuff...
   $app.dao().saveRecord(record)
}

$app.dao().runInTransaction(( txDao )=>{
   foo()
})
```

**Comment by ganigeorgiev** [comment, marked as answer]
There will be a deadlock. You need to pass the `txDao` to your function.

**Reply by benallfree** [reply]
When are `onModel*` handlers triggered if records are saved inside a transaction? Could two handlers cause a deadlock?

**Reply by ganigeorgiev** [reply]
> When are onModel* handlers triggered if records are saved inside a transaction?

The `before*` model hooks will fire immediately when the write operation is invoked.
The `after*` model hooks fire after the transaction was committed successfully.

> Could two handlers cause a deadlock?

It shouldn't, they will just wait their turn. The deadlock in the above happens because the transaction cannot proceed further because it will wait the write `app.Dao()` operation to complete, which in turn cannot also proceed because it needs to wait the transaction.

**Reply by benallfree** [reply]
I see that dao is optionally past into the onModel hook. Under what conditions does that happen?

**Reply by ganigeorgiev** [reply]
It should be always there, it is not optional.
If you are referring to the JSVM event type definition, this is because pointer Go values could be `nil`.

**Comment by benallfree** [comment]
Is there any way to detect this condition? Maybe hook timeouts, or a way to query the execution pool to detect if it is running low?

**Reply by ganigeorgiev** [reply]
From userland code I'm not sure.
It definitely could be handled better and more user friendly, eg. return an error with message that it will deadlock (_transactions and the connections pool will be revisited sometime in the future in regards to the planned bulk web APIs and the option to wrap in a transaction `*Request` before hooks_).

**Reply by ghost** [reply]
I would love if this condition was detected / avoided by pb.
